mapper/orm/src/main/java/org/hibernate/search/mapper/orm/search/loading/impl/EntityLoaderBuilder.java

```diff
@@ -8,50 +8,79 @@
 
 import java.util.HashMap;
 import java.util.Map;
-import java.util.Set;
+
+import javax.persistence.metamodel.IdentifiableType;
+import javax.persistence.metamodel.SingularAttribute;
 
 import org.hibernate.Session;
 import org.hibernate.search.engine.search.loading.spi.EntityLoader;
+import org.hibernate.search.mapper.pojo.mapping.spi.PojoMappingTypeMetadata;
 import org.hibernate.search.mapper.pojo.search.PojoReference;
 
 public class EntityLoaderBuilder<E> {
 
 	private final Session session;
-	private final Set<Class<? extends E>> concreteIndexedClasses;
+	private final Map<Class<? extends E>, PojoMappingTypeMetadata> concreteIndexedClassesToMetadata;
 
-	public EntityLoaderBuilder(Session session, Set<Class<? extends E>> concreteIndexedClasses) {
+	public EntityLoaderBuilder(Session session,
+			Map<Class<? extends E>, PojoMappingTypeMetadata> concreteIndexedClassesToMetadata) {
 		this.session = session;
-		this.concreteIndexedClasses = concreteIndexedClasses;
+		this.concreteIndexedClassesToMetadata = concreteIndexedClassesToMetadata;
 	}
 
-	public EntityLoader<PojoReference, E> build(MutableEntityLoadingOptions mutableLoadingOptions) {
-		if ( concreteIndexedClasses.size() == 1 ) {
-			Class<? extends E> concreteIndexedType = concreteIndexedClasses.iterator().next();
-			return buildForSingleType( mutableLoadingOptions, concreteIndexedType );
+	public EntityLoader<PojoReference, ? extends E> build(MutableEntityLoadingOptions mutableLoadingOptions) {
+		if ( concreteIndexedClassesToMetadata.size() == 1 ) {
+			Map.Entry<Class<? extends E>, PojoMappingTypeMetadata> entry =
+					concreteIndexedClassesToMetadata.entrySet().iterator().next();
+			return buildForSingleType( mutableLoadingOptions, entry.getKey(), entry.getValue() );
 		}
 		else {
 			return buildForMultipleTypes( mutableLoadingOptions );
 		}
 	}
 
-	private HibernateOrmComposableEntityLoader<PojoReference, E> buildForSingleType(
-			MutableEntityLoadingOptions mutableLoadingOptions, Class<? extends E> concreteIndexedType) {
+	private <E2 extends E> HibernateOrmComposableEntityLoader<PojoReference, E2> buildForSingleType(
+			MutableEntityLoadingOptions mutableLoadingOptions,
+			Class<E2> concreteIndexedType, PojoMappingTypeMetadata metadata) {
+		// TODO HSEARCH-3349 Add support for other types of database retrieval and object lookup?
+		//  See HSearch 5: org.hibernate.search.engine.query.hibernate.impl.EntityLoaderBuilder#getObjectInitializer
+
+		if ( metadata.isDocumentIdMappedToEntityId() ) {
+			return new HibernateOrmSingleTypeByIdEntityLoader<>(
+					session,
+					concreteIndexedType,
+					mutableLoadingOptions
+			);
+		}
+		else {
+			IdentifiableType<E2> indexTypeModel = session.getSessionFactory().getMetamodel().entity( concreteIndexedType );
+			SingularAttribute<? super E2, ?> documentIdSourceProperty =
+					indexTypeModel.getSingularAttribute( metadata.getDocumentIdSourcePropertyName().get() );
```

Comment:
What does it happen if `metadata.getDocumentIdSourcePropertyName()` is not present here?

Reply:
It will not happen with the ORM mapper:

* The POJO mapper throws an exception when the document ID is not mapped and "provided IDs" are disabled
* The ORM mapper disables provided IDs.